Floating_IPs: Use API Get + Cloud_Network helper

gildub opened the pull request:
Networks > Floating IPs

* Replaces GET routes/actions by API calls (https://github.com/ManageIQ/manageiq-ui-classic/issues/737)
* Consolidates edit and new views
* Adds getCloudNetworksByEms helper to be used by others network features.
* Description field is gone, although the field is present on OpenStack side but in MIQ Model.
* Fixed IP address cannot be edited

The tenant placement field has been moved up on the top of the form in order to be consistent with other network forms:
* Before this patch:
<img width="1193" alt="before" src="https://user-images.githubusercontent.com/1901741/32868318-65bad462-cabd-11e7-932d-8601e80a7dd0.png">

* After this patch:
<img width="1198" alt="after-tenant-placement" src="https://user-images.githubusercontent.com/1901741/32868335-7fca1872-cabd-11e7-977f-5756eb0efb87.png">

The edit form (associate/disassociate floating IP) display other fields as well:
* Before this patch:
<img width="1196" alt="before-edit" src="https://user-images.githubusercontent.com/1901741/32868373-b51625ca-cabd-11e7-9252-1f43dbeea54f.png">

* After this patch:
<img width="1194" alt="after-edit" src="https://user-images.githubusercontent.com/1901741/32868389-c854e360-cabd-11e7-9671-e952e315f4bf.png">

gildub commented:
More PRs to follow for the controller to component' and then move the rest of the http requests (create/update/delete) to API only.

@martinpovolny, @himdel, @ZitaNemeckova

gildub commented:
@aufi, do you happen to know what's the involved CI error?

aufi commented:
@gildub I'm not sure, but the error looks relevant. Propably caused by some recent UI change or this PR..

gildub commented:
@aufi, you're right, ok it's fixed.
...You know, with all those Travis drama we had lately...
Thanks.

gildub commented:
@aufi, actually something went wrong, back to [WIP].

gildub commented:
@martinpovolny, @himdel, could you please assign someone for review?

gildub commented:
@dclarizio, would you mind assigning this one please?

ZitaNemeckova commented on app/assets/javascripts/controllers/floating_ip/floating_ip_form_controller.js:
```diff
@@ -1,34 +1,42 @@
+    vm.saveable = miqService.saveable;
+
+    if (vm.newRecord) {
+      vm.afterGet = true;
```
This should be in `.then` like it is in `else` below. Otherwise form will be displayed before API call is resolved.

gildub replied:
You're right I forgot that flag triggers the form 👍. Done

ZitaNemeckova commented on app/assets/javascripts/controllers/floating_ip/floating_ip_form_controller.js:
```diff
@@ -42,34 +50,18 @@ ManageIQ.angular.app.controller('floatingIpFormController', ['$http', '$scope',
-      .then(getNetworkByEmsFormData)
-      .catch(miqService.handleFailure);
-
+    miqService.getCloudNetworksByEms(function(data) {
```
Why did you move API call to `miqService`? If it's not something common for multiple controllers it should stay in its own controller.

gildub replied:
@ZitaNemeckova, I'm not following you, the `getCloudNetworksByEms` which is now in `miq_service.js` is using both a Promise [1], the `then` part, where the promise is the callback itself.

[1] https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/assets/javascripts/miq_api.js#L5

ZitaNemeckova replied:
Yeah, I suggested to use Promise instead of callback but that's out of the scope for this PR. And this only populates fields after a change there's no need for returning Promise. Please ignore that comment :)

ZitaNemeckova commented:
@gildub Please add Before/After screenshots in PR description as `Cloud Tenant` has been moved and some fields changed names and one is no longer disabled. Otherwise it's hard to tell if differences below are mistakes or not. 

<img width="1193" alt="screen shot 2017-11-13 at 11 29 00 am" src="https://user-images.githubusercontent.com/9210860/32721156-e53407e0-c865-11e7-9158-7c01e8c27868.png">
<img width="1198" alt="screen shot 2017-11-13 at 11 21 52 am" src="https://user-images.githubusercontent.com/9210860/32720969-50139b94-c865-11e7-960e-caadfa2cb251.png">

ZitaNemeckova commented on app/controllers/floating_ip_controller.rb:
```diff
@@ -128,51 +128,15 @@ def edit
-      :url  => "/floating_ip/new"
-    )
+    drop_breadcrumb(:name => _("Add New Floating IP"), :url => "/floating_ip/new")
+    render :change
```
I never saw this before. Why not use `render :partial =>... ` or `render :template =>`?

gildub replied:
The consolidation of new and edit forms ended up not using a partial therefore it `render` directly the consolidated form which I named `change`. I believe it's neither good or bad, just a name and definitely not worse than `form`, which like calling an array variable `array`.

End of the day this doesn't matter because the move from controller to component t (see https://github.com/ManageIQ/manageiq-ui-classic/pull/2530) this is going to disappear.

In the meantime I've renamed it to `new_edit.html.haml`.

ZitaNemeckova commented on app/controllers/floating_ip_controller.rb:
```diff
@@ -128,15 +128,15 @@ def edit
       :name => _("Associate Floating IP \"%{address}\"") % { :address => @floating_ip.address },
       :url  => "/floating_ip/edit/#{@floating_ip.id}")
-    render :change
+    render :new_edit
```
This works but is inconsistent with usual way it's done. It should be `render :action => "floating_ips/new_edit"` or `render :template => "floating_ips/new_edit"`. So my main concern is consistency :)

gildub replied:
Sure. Done

gildub replied:
I disagree with this, the RoR controller provides the context this is why there is no need for absolute path and as a matter of fact, there is no other example across the GUI besides the special case of "layouts". So unless there is a new "practice" to be set, I'm keeping it this way. On top of that, as I already mentioned before, this doesn't matter since the component approach takes over the rendering.

ZitaNemeckova replied:
I agree that it works :) but it's **never** used this way in the project (except once by you). Common practice here would be to have `new` and `edit` templates (that will be rendered automatically) that have something like
``` haml
#main_div
  = render :partial => 'floating_ip_form'
```
and rename `new_edit` to `_floating_ip_form`.

Or if that is not an option then `render :action => "new_edit"` or `render :template => "new_edit"` (error in first comment) to be consistent with practice in this project.

Please choose one :)

ZitaNemeckova commented on app/assets/javascripts/controllers/floating_ip/floating_ip_form_controller.js:
```diff
@@ -1,34 +1,43 @@
 
   vm.addClicked = function() {
-    var url = 'create/new' + '?button=add';
+    var url = 'create/new?button=add';
```
This should be `'/floating_ip/create/new?button=add'`  as below in `vm.cancelClicked`. That way you don't assume it's gonna be `/floating_ip/` + `url`.

gildub replied:
Sure.

himdel commented on app/controllers/floating_ip_controller.rb:
```diff
@@ -128,51 +128,15 @@ def edit
-      :url  => "/floating_ip/new"
-    )
+    drop_breadcrumb(:name => _("Add New Floating IP"), :url => "/floating_ip/new")
+    render :new_edit
```
(and here)

gildub replied:
Done too.

himdel replied:
This is still not fixed, and neither is the other occurence on [line 131](https://github.com/ManageIQ/manageiq-ui-classic/pull/2476/files#diff-cb580b24d878da0531aed3049acfe8d4R131).

Use explicit `render :action => ...` or `render :template => ...` or create the 2 templates.

gildub commented:
@ZitaNemeckova, @himdel, cleared issues, thanks

himdel commented:
Trying to submit a new item:

before:
`description:`
`floating_ip_address:1.2.3.4`
`network_port_ems_ref:123`

after:
`address:1.2.3.4`
`network_port[ems_ref]:123`
`fixed_ip_address:5.6.7.8`

Looks like the code is still looking for `floating_ip_address`, not `address`, and for `network_port_ems_ref`, not the other one:

```ruby
    options[:floating_ip_address] = params[:floating_ip_address] if params[:floating_ip_address]
    ...
    options[:network_port_ems_ref] = params[:network_port_ems_ref] if params[:network_port_ems_ref]
```

Also, there's no code looking at `fixed_ip_address` and the field used to be read-only - a bug?

(`description` is indeed dead).

himdel commented:
Also, the list of managers now includes all network mangers, no longer filtered by Openstack.

gildub commented:
@himdel, 

* Description field is gone, that was a mistake because the field exists on OpenStack side but not within MIQ Model:

* Fixed IP address cannot be edited
I updated description accordingly for those two.

Regarding the list of managers, it's expected to get all the network managers and not just the OpenStack one, as we're dealing with network objects the user to select the desired one. The supported feature are then managed on a per provider basis.

> Looks like the code is still looking for floating_ip_address, not address, and for network_port_ems_ref...

There was a mismatch that I've addressed, beyond that I'm not sure what you mean, do you see a specific error?

gildub commented:
@himdel, could you please provide update/follow-up? Thanks

himdel commented:
Still unadressed: https://github.com/ManageIQ/manageiq-ui-classic/pull/2476#discussion_r153491149

Also, https://github.com/ManageIQ/manageiq-ui-classic/pull/2476#issuecomment-347522487 doesn't appear adressed either, you're still sending the data under a different name than the backend is using.

gildub commented:
@himdel, update please

gildub commented:
@ZitaNemeckova, @himdel, please review.

himdel commented:
Rebased,
and hopefully resolved https://github.com/ManageIQ/manageiq-ui-classic/pull/2476#discussion_r150517590, https://github.com/ManageIQ/manageiq-ui-classic/pull/2476#discussion_r153156900 and https://github.com/ManageIQ/manageiq-ui-classic/pull/2476#discussion_r153491149 by removing the shared "action" in favor of how this is done everywhere else.

Will continue with the review...

himdel commented:
Verified that `self.raw_create_floating_ip` expects

* `cloud_network_id`
* `cloud_tenant_id`
* `address`
* `network_port_ems_ref`

(and `ems_id` is deleted on the way) .. so `form_params` output still looks good.

This matches `vm.fields` now, so we should not be sending anything else :+1:

himdel commented:
OK, so this should now be ready to merge from my perspective .. @gildub , @aufi can you make sure this is still working as intended?

I took the liberty of rebasing, cleaning up some rough edges (see commits :)), and sorting out the list of fields we're sending and letting the user edit:

* removed `router_id` completely, this was never in the form or used
* made `fixed_ip_address` read only again, this is never saved anywhere (but it is shown in textual summary so it does make sense to display)
* unhidden the `address` field in edit - looks like this was only meant to be disabled, not hidden


Obligatory before and after now:

`/new` - master
![new old](https://user-images.githubusercontent.com/289743/37227989-fef7bd12-23d6-11e8-943e-3a9096af03d0.png)

`/new` - now
![new new](https://user-images.githubusercontent.com/289743/37227992-0074c504-23d7-11e8-9c0b-8ed1c0f09522.png)

`/edit` - master
![edit old](https://user-images.githubusercontent.com/289743/37228002-08967804-23d7-11e8-838b-c6301c65a487.png)

`/edit` - now
![edit new](https://user-images.githubusercontent.com/289743/37228009-0b4d7886-23d7-11e8-9294-87f1e039d327.png)

martinpovolny commented:
Ping  @gildub, @aufi, @tzumainn

gildub commented:
@himdel, @martinpovolny, I rebased the modified remote branch, and retested, LGTM, thanks.

himdel commented:
@gildub You also managed to remove all of my changes when rebasing :). So .. all of my original review still remains unresolved ;).

I'm ..not risking it again, so I'm pushing the changes as a separate PR, please test that one instead :).

gildub commented:
I love the "also" part. Since you didn't provided any details, I pull my remote branch branch which you updated, seems like that didn't work.
